### ISSUE TITLE
Allow EDD Session to start in wp-admin

### ISSUE THREAD

**Issue by mintplugins** [easydigitaldownloads/easy-digital-downloads#6648]
## Bug Report
The EDD Session was not allowed to start in #6240, but that is causing files uploaded by Vendors in Frontend Submissions to not be placed in the protected `edd` directory. See this FES issue for details about that:
https://github.com/easydigitaldownloads/edd-fes/issues/1408 

We can fix this by taking out the edd session prevention in wp-admin, while still allowing the session to be prevented just for WP Core file editor saves. This will leave #6240 resolved, while also fixing the issue in FES. 

### Steps to reproduce the behavior and fix
1) On EDD `Master`, in FES try uploading a file as a Vendor. It will not go into the protected `edd` directory. 
2) Switch to EDD branch `issue/6648` and try uploading a file as a Vendor in FES again. This time it should go into the protected `edd` directory. 
3) To make sure the fix is still working for #6240, also try saving a theme php file under "Dashboard" > "Appearance" > "Editor". It should save without any issues.

**Comment by mintplugins** [comment]
@SDavisMedia @pippinsplugins If you guys can take this for a quick test when available that would be great. Thanks!

**Comment by pippinsplugins** [comment]
Looks good here!

**Comment by chriscct7** [comment]
Reopening. This caused a test server of ours to get 24 million rows because a lot of things run in the admin. 

I personally have never seen another usecase for admin EDD sessions.

If FES needs this, I would suggest strongly reverting this change, then in FES have it use the existing filter `edd_start_session` and if start = false, then if admin && current page is a download then do it. That way only sessions are started on the download screen and only for those using FES.

**Comment by chriscct7** [comment]
We were able to revert this change via filters but others who use plugins that have admin based endpoints will not be as lucky

**Comment by mintplugins** [comment]
@chriscct7 Can you give us a bit more info? Which rows were created? Was it custom code that caused the rows, or was it core code activated only? Which filters are you referring to that you used to revert it? As much detail as possible will definitely help.

**Comment by mintplugins** [comment]
After following up with @chriscct7 on Slack, the rows being created are the `_wp_session_` transient rows, which are added to the options table as part of the WP_Session class, which is included in EDD here:

https://github.com/easydigitaldownloads/easy-digital-downloads/blob/986c1ff1c7dc856ba005892dbfbadfbd95bc995f/includes/class-edd-session.php#L94

The reason there are millions being created is because of a backup plugin which hits the site millions of times per day, and creating a new session each time, which creates a new transient each time. Something seems wrong with that, as it should either be cleaning up its sessions when it leaves through setting transient expirations via the `wp_session_expiration_variant` and `wp_session_expiration` filters, or using the same session it previously did, depending on what it is doing. 

Regardless, we'll take a look at removing the session from wp-admin again. However, this time we must do a complete investigation of all extensions to ensure none rely on on the EDD Session in wp-admin. This is something we must do to help prevent what happened with Frontend Submissions, where [Vendor uploads were uploaded to an unprotected directory for 6 months](https://github.com/easydigitaldownloads/edd-fes/issues/1408).

**Comment by chriscct7** [comment]
What should be done here is simple. The change to EDD core should be reverted.

Then FES, which is the only extension that needs this, should use the existing hook:
```
add_filter( 'edd_start_session', 'fes_custom_start_sessions' );
function fes_custom_start_sessions( $start_session ) {
   // if we're not planning on starting sessions, then see if we're in the admin
    if( ! $start_session && is_admin() ) {
		$start_session = true;
	}
    return $start_session;
}
```

That way the sessions are only started in the admin for FES users. 

Further, which the above is written simply to undo what was done, this condition of turning it on could be further reduced by seeing if we're on the new or edit download screen through edd_is_admin_page.

**Comment by chriscct7** [comment]
PR for EDD core: https://github.com/easydigitaldownloads/easy-digital-downloads/pull/6864
PR for FES core: https://github.com/easydigitaldownloads/edd-fes/pull/1413

**Comment by JJJ** [comment]
Reverted from EDD core and merged both PRs, completing the move to FES.

Thanks @chriscct7 👍

**Comment by mintplugins** [comment]
Just to clarify - prior to merging this, were all other extensions investigated to ensure they do not rely on edd session in wp-admin?

**Comment by JJJ** [comment]
>were all other extensions investigated

This level of depth is not a documented requirement for any other pull request.

If it is going forward, who is writing the integration tests? It's an unreasonable ask, to have _everyone_ install a hundred extensions and test them all for every pull request.

>If so, that should be communicated clearly here

I imagine when @pippinsplugins said "Looks good here!" he did not take the next 20 hours to install all other extensions and run every function to guarantee nothing broke.

I personally tested the changes that @chriscct7 proposed _here_, confirming they correctly fix the issue as its reported _here_. His fix is correct, yours broke at least one large site.

>If not, that needs to happen prior to a change like this, as it caused major problems for our customers previously

This change broke sites now, and it likely broke others that haven't noticed yet because most folks don't go kicking around their `wp_options` table looking for hidden session keys.

We were lucky enough to have someone familiar with contributing to open source software report it and offer to help in the form of research and a pull-request.

Reverting the original change in favor of one that addresses FES as documented in this issue, seems like the very obvious thing to do.

If there's historical reference that's critical to your issue, including that here with words is far more helpful than linking to another issue and hoping folks will come to the same conclusion that you did. 

This issue as reported, ironically enough, omits the most important parts of the "New Issue" template that's setup here. `Expected behavior` `Actual behavior` `Information (if a specific version is affected)` would have been useful.

----

His PR addresses your concerns in way that doesn't involve adding sessions to every single `wp-admin` request, which is certainly a mistake - it is a guaranteed way to destroy any caching that most web hosts employ to keep `wp-admin` speedy. `admin-ajax.php` is allowed because its already sending `nocache_headers()`, and hosts can plan for that. Not caching all of 

----

All of this is to say that I don't think any of this matters to the original problem, which is that the upload path isn't correct under certain conditions. Storing the upload path client side is a terrible idea as it's totally untrustworthy. If someone modifies the session so the upload path is `../../../etc/` do you really want to trust that is actually correct?

**Comment by pippinsplugins** [comment]
I don't think @mintplugins's `were all other extensions investigated to ensure they do not rely on edd session in wp-admin` was meant to suggest that hundreds of extensions be tested, just the ones we know of that do rely on sessions in the admin (aka anything with ajax uploads). That's FES and only FES as far as our own extensions go.

**Comment by mintplugins** [comment]
@pippinsplugins That's what I was trying to say, yes. My apologies for not being more clear @JJJ.

I believe I've found one other place where the session might be used in admin in EDD core:
https://github.com/easydigitaldownloads/easy-digital-downloads/blob/c38c0820f18a472ceba6396e12cbcd74743227e5/includes/admin/settings/register-settings.php#L1164

**Comment by pippinsplugins** [comment]
Aha, yes! @mintplugins you are correct. We do use sessions in the admin when enabling previously-disabling Sequential Order Numbers.

**Comment by chriscct7** [comment]
@pippinsplugins that's a usecase that also gets discontinued with EDD 3.0 though right as the orders are now static post-completion

**Comment by pippinsplugins** [comment]
I believe you are right @chriscct7
